tests/reactor_test.py

```diff
@@ -0,0 +1,254 @@
+import os
+import socket
+import threading
+import time
+from collections import OrderedDict
+
+from mock import MagicMock
+from parameterized import parameterized
+
+from hazelcast import six
+from hazelcast.reactor import AsyncoreReactor, _WakeableLoop, _SocketedWaker, _PipedWaker, _BasicLoop
+from hazelcast.util import AtomicInteger
+from tests.base import HazelcastTestCase
+
+
+class ReactorTest(HazelcastTestCase):
+    def test_default_loop_is_wakeable(self):
+        reactor = AsyncoreReactor()
+        self.assertIsInstance(reactor._loop, _WakeableLoop)
+
+    def test_reactor_lifetime(self):
+        t_count = threading.active_count()
+        reactor = AsyncoreReactor()
+        reactor.start()
+        try:
+            self.assertEqual(t_count + 1, threading.active_count())  # reactor thread
+        finally:
+            reactor.shutdown()
+        self.assertEqual(t_count, threading.active_count())
+
+
+LOOP_CLASSES = [
+    ("wakeable", _WakeableLoop,),
+    ("basic", _BasicLoop,),
+]
+
+
+class LoopTest(HazelcastTestCase):
+    def test_wakeable_loop_default_waker(self):
+        loop = _WakeableLoop({})
+        try:
+            if os.name == "nt":
+                self.assertIsInstance(loop.waker, _SocketedWaker)
+            else:
+                self.assertIsInstance(loop.waker, _PipedWaker)
+        finally:
+            loop.waker.close()
+
+    def test_wakeable_loop_waker_closes_last(self):
+        dispatchers = OrderedDict()
+        loop = _WakeableLoop(dispatchers)  # Waker comes first in the dict
+
+        mock_dispatcher = MagicMock(readable=lambda: False, writeable=lambda: False)
+        dispatchers[loop.waker._fileno + 1] = mock_dispatcher
+
+        original_close = loop.waker.close
+
+        def assertion():
+            mock_dispatcher.close.assert_called()
+            original_close()
+
+        loop.waker.close = assertion
+
+        loop.shutdown()
+
+    @parameterized.expand(LOOP_CLASSES)
+    def test_check_loop(self, _, cls):
+        loop = cls({})
+        loop.check_loop()
```

Comment:
nit: I'd leave a comment that we rely on `assert`s placed in the method implementation here to make things a bit more straight-forward for future readers.

Reply:
Wrote a comment explaining that